### PR TITLE
DBT: add branch creation capability

### LINKED ISSUE #2928: create-branch-schema -  create lakefs branch
create-branch-schema command should have an option to create the branch in lakefs

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
Currently when users want to copy their dbt schema from one branch to another they need to create this branch and then use the `lakectl dbt create-branch-schema` to copy the schema to the dedicated branch created.
This PR introduces the `--create-branch` flag which enables the user to create the branch in the context of `lakectl dbt create-branch-schema`. Since a schema is defined in the dbt project and is associated with a single branch, the source branch (from which we diverge) is the same as the schema's. 

Closes #2928

### CODE REVIEW

File: cmd/lakectl/cmd/branch_tools.go
Comment (idanovo) [inline cmd/lakectl/cmd/branch_tools.go:6]: You can use `goimports` command in order to fix imports order issues

File: cmd/lakectl/cmd/branch_tools.go
Comment (idanovo) [inline cmd/lakectl/cmd/branch_tools.go:38]: If you merge this after I merge [this](https://github.com/treeverse/lakeFS/pull/2966/files), please change this line to be `DieOnErrorOrUnexpectedStatusCode(resp, err, http.StatusCreated)`

File: cmd/lakectl/cmd/dbt.go
Comment (idanovo) [inline cmd/lakectl/cmd/dbt.go:52]: Why not use `DieErr` in this case?

File: cmd/lakectl/cmd/dbt.go
Comment (idanovo) [inline cmd/lakectl/cmd/dbt.go:74]:
```suggestion
	fmt.Println("dbt debug succeeded with schemaOutput:", schemaOutput)
```

File: cmd/lakectl/cmd/dbt.go
Comment (Jonathan-Rosenberg) [reply cmd/lakectl/cmd/dbt.go:53]: As I understand it, `DieFmt` is used for invalid input. 
In this case the given schema identifier (in the dbt macro) is not the expected `lakectlIdentifier` identifier given here. Although it's not a "classic" input it does seem fit..

**Review by idanovo — COMMENTED** [review]
Looks good! Only the comment about the `DieOnResponseError` is important

File: cmd/lakectl/cmd/branch_tools.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/branch_tools.go:15]:
```suggestion
		return "", "", fmt.Errorf("\get database on extractBranchFromSchema from '%s': %w", dbName, err)
```
(Go errors describe what happened but do not say it is bad)

Also below.

File: cmd/lakectl/cmd/branch_tools.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/branch_tools.go:40]: Please use a formatting output with a template.  It supports colouring and other neat features, and is generally simpler to maintain (for user-visible outputs).

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/branch_tools_test.go:37]: Nit:
```suggestion
			name: "Valid",
```
(or similar).  The other test flows are less happy, but I don't want to say "Sad" about each of them.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/branch_tools_test.go:25]: If we never pass a context, can we just have `dbName` here?  (If we _do_ pass a context, Go generally prefers `context.Background()` or `context.TODO()` to the `nil` context...).

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/branch_tools_test.go:33]: This test is very brittle.

Can `wantErr` be an `error`, that we can then check with `errors.Is`?  Or, if the is impossible, a substring of the expected error??  Otherwise any scenario with `wantErr == true` (most of them) will happily pass as long as _something_ on the execution path fails.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/branch_tools_test.go:32]: Please document these fields.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/branch_tools_test.go:34]: Please export all fields in a `tests` struct.  It is often tempting to use`deep.Equal` or something else that relies on reflection on these at some later stage.  And that will compare no fields because Go reflection cannot see lowercase (unexported) fields.

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt.go:22]:
```suggestion
	// Debug validates dbt connection using dbt debug and returns the schema
	// configured by the target environment (configured in the dbt profiles file)
```

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt.go:27]: Consider documenting that it dies if something is invalid.

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt.go:19]:
```suggestion
type DBTCommandClient interface {
```
(and in other `type`s and `func`s below): the Go way is to keep capitalization of acronyms and abbreviations.  So we always have `XyzzyID` (with a capital D), etc.

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [reply cmd/lakectl/cmd/dbt.go:53]: `DieErr` is the way to die on an error.  `DieFmt` is if you have a message but not an error.  (Yes, one could be implemented in terms of the other.)

In any case, `DieFmt` is dangerous to use like this because the error message may contain formatting chars.  If you must, use `DieFmt("%s", ...)`.

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt.go:111]:
```suggestion
		if toSchema == "" {
```

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt.go:117]:
```suggestion
		metastoreClient, clos := getMetastoreClient(clientType, "")
		defer close()
```
(name the variable for what it is, not for how it will be used on the next line).

**Review by arielshaqed — COMMENTED** [review]
Thanks, looks really good!

Added many Go style comments (sorry).  Not enough of an expert on DBT to say much interesting on that side, except for... "how was this tested?".

Please also fix Go linter warnings.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (Jonathan-Rosenberg) [reply cmd/lakectl/cmd/branch_tools_test.go:37]: The other flows are sad from my POV but sure 😉

File: cmd/lakectl/cmd/branch_tools.go
Comment (johnnyaug) [inline cmd/lakectl/cmd/branch_tools.go:85]: You had the repository and branch names in hand - then you made a URI out of them, and parsed the URI again to extract the names.

Why not just use the names in the first place?

File: pkg/dbtutil/dbt_util.go
Comment (johnnyaug) [inline pkg/dbtutil/dbt_util.go:1]: IMO this file doesn't justify a top-level package.
Maybe it can go under metastore > dbt > utils.go.
@nopcoder may have an opinion.

File: pkg/dbtutil/dbt_util_test.go
Comment (johnnyaug) [inline pkg/dbtutil/dbt_util_test.go:277]:
```suggestion
func toJSON(resources ...DBTResource) string {
```

File: cmd/lakectl/cmd/branch_tools.go
Comment (johnnyaug) [inline cmd/lakectl/cmd/branch_tools.go:23]: That's a lot of work to create custom errors - and I'm not sure why they are needed.

Can something like the following work?
```
fmt.Errorf("trying to reach database: %w", originalErr)
```

**Review by johnnyaug — CHANGES_REQUESTED** [review]
Looking good! And totally makes sense to implement.

I think some parts can be shorter and some naming can be improved - left comments.

File: cmd/lakectl/cmd/branch_tools.go
Comment (Jonathan-Rosenberg) [reply cmd/lakectl/cmd/branch_tools.go:72]: As a general purpose method (not part of a business logic flow) I don't think we can assume anything about the input...
WDYT?

Comment (johnnyaug) [reply cmd/lakectl/cmd/branch_tools.go:72]: I think it's fine to assume that the arguments are valid at this point

File: cmd/lakectl/cmd/branch_tools.go
Comment (johnnyaug) [reply cmd/lakectl/cmd/branch_tools.go:85]: What I mean is, why do you need a URI at all?

Comment (Jonathan-Rosenberg) [reply cmd/lakectl/cmd/branch_tools.go:85]: The purpose was to create a function that takes URI strings and builds a new branch out of them, that way if other flows will need to create a branch from given URI strings it would be easier for them (they won't need to parse the URIs but just pass it to this function).
I could have created a function that takes the repo and branch as you said but I thought it would be more useful to receive the URI strings because the convention is that the user inputs the full lakefs path `lakefs://{repo}/{branch}` rather then the repo and branch separately, thus making the use of this function easier...

File: cmd/lakectl/cmd/branch_tools.go
Comment (Jonathan-Rosenberg) [reply cmd/lakectl/cmd/branch_tools.go:23]: This was done to test the code not by the string but by the type of the error...
Go doesn't consider custom errors to be a good practice? (truly don't know 😅 )

File: test/lakectl_metastore/dbt-entrypoint.sh
Comment (Jonathan-Rosenberg) [reply test/lakectl_metastore/dbt-entrypoint.sh:17]: If we take it off then we'll need to `lakectl commit` the previous step (`lakectl dbt create-branch-schema ..`) so the next will succeed. 
Why is it diverting our attention?

File: cmd/lakectl/cmd/common_helpers.go
Comment (Jonathan-Rosenberg) [reply cmd/lakectl/cmd/common_helpers.go:204]: It might be a handy function but as I see it, no method is using it expect for `DieOnErrorOrUnexpectedStatusCode`, so what are the odds the a new one will?
On top of that, developers that used to use `DieOnResponseError` might use it again as a force of habit which we want to prevent.
wdyt?

Comment (idanovo) [reply cmd/lakectl/cmd/common_helpers.go:204]: Agree with @Jonathan-Rosenberg about this. we want to prevent users from using `DieOnResponseError`

File: cmd/lakectl/cmd/branch_tools.go
Comment (johnnyaug) [reply cmd/lakectl/cmd/branch_tools.go:80]: I think we should not generate lakeFS URIs in our code.
This is error prone. URIs should be used from the outside to identify resources in lakeFS.

Also, if you choose to stay with this function it may belong in the "uri" package.

**Comment by johnnyaug** [comment]
Nice! I think it's close to being ready.
Not approving because there is one open discussion about whether to generate URIs in the code, and also one unaddressed comment about the creation of the "dbt_utils" package.

File: cmd/lakectl/cmd/branch_tools.go
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_tools.go:69]: Looks like internal func and not public.
Name suggestion - use Format... and not Generate.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_tools_test.go:117]: Testing the implementing by the code the same implementation - this is usually the point where I stop and explain to myself why and what I want to write test code in the first place.
In this case, I don't see the point in testing Sprintf and use a validation function or working code that use the data as an input to validate the value is what I expected - without re-implement it.

File: cmd/lakectl/cmd/branch_tools.go
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_tools.go:75]: I think in this level, you should accept the parsed version and not `MustXXX`. Calling `MustParseRefURI` is handling user input or constants not our generated input.
In this case, it looks like we could just accept each repository and branch name.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_tools_test.go:21]: Can be added to the constants above just by using '+'. But I really prefer to see these values inside the single func scope it is used.  Using globals, even for test code, without a specifc name - it is something I try to avoid.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_tools_test.go:78]: Why not passing the correct constant in the test data? It will also open the case where we like to pass something else like empty string.

File: cmd/lakectl/cmd/branch_tools_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_tools_test.go:95]: Naming: looks like NewMockClient

File: cmd/lakectl/cmd/dbt.go
Comment (Jonathan-Rosenberg) [reply cmd/lakectl/cmd/dbt.go:48]: 1. Will be changed.
2. Why don't you consider that an error? If the user asks for a branch to be created and a schema be created with relation to it, the user needs its current dbt defined schema to be initialized with a branch location. If the user asks for a branch to get created and there's no such schema, it's an error just like if the user would want to create a branch with a source branch that doesn't exist, no?

File: cmd/lakectl/cmd/dbt_util_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/dbt_util_test.go:245]:
```suggestion
				t.Errorf("DbtRun() error = '%v', expected error = '%v'", err, tt.Err)
```

File: cmd/lakectl/cmd/dbt_util_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/dbt_util_test.go:249]:
```suggestion
				t.Errorf("DbtRun() runOutput = %v, expected = %v", runOutput, tt.RunOutput)
```

File: cmd/lakectl/cmd/dbt_util_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/dbt_util_test.go:227]: when every arg is the same for all tests I usually remove it from the code and just pass it as argument

File: cmd/lakectl/cmd/dbt_util_test.go
Comment (nopcoder) [inline cmd/lakectl/cmd/dbt_util_test.go:24]: can we use:
```go
var errDummyCommandFailed = errors.New("boom!")
```

**Review by nopcoder — APPROVED** [review]
looks good - added some suggestions and alignments.

File: cmd/lakectl/cmd/dbt_util.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt_util.go:38]: I don't understand this error message, specifically "Handle lakefs support".

In general error messages should not contain more than a single phrase, because `Errorf` will be used on them.

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt.go:111]: Another nit if we're already here: style is "DBT" not "Dbt".

```suggestion
func copySchemaWithDBTTables(ctx context.Context, continueOnError, continueOnSchemaExists bool, fromSchema, toSchema, branchName, dbfsLocation, projectRoot string, client metastore.Client) {
```

File: cmd/lakectl/cmd/dbt.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/dbt.go:131]: More nits:
```suggestion
	output, err := DBTRun(projectRoot, toSchema, lakectlIdentifier, []string{"config.materialized:view"}, commandExecutioner{})
```
(Can you go over the entire file and see we have no "Dbt", please?  (Can be in another PR).

**Review by arielshaqed — APPROVED** [review]
Thanks!

Sorry for the lengthy review.  Approving; please see if you can:
1. Improve strange error message.
2. Rename `commandExecutor` (many comments).
3. Get rid of "Dbt" in favour of "DBT" (can be in a separate PR).

**Review by johnnyaug — APPROVED** [review]
Looks very good!
Thanks for addressing all the comments!

**Pull request merged**
